diff --git a/conf/cassandra-env.sh b/conf/cassandra-env.sh
index a640847..8d972f7 100644
--- a/conf/cassandra-env.sh
+++ b/conf/cassandra-env.sh
@@ -166,8 +166,10 @@ USING_G1=$?
 # times. If in doubt, and if you do not particularly want to tweak, go with
 # 100 MB per physical CPU core.
 
-#MAX_HEAP_SIZE="4G"
-#HEAP_NEWSIZE="800M"
+MAX_HEAP_SIZE="500M"
+HEAP_NEWSIZE="100M"
+#MAX_HEAP_SIZE="1000M"
+#HEAP_NEWSIZE="200M"
 
 # Set this to control the amount of arenas per-thread in glibc
 #export MALLOC_ARENA_MAX=4
diff --git a/conf/cassandra.yaml b/conf/cassandra.yaml
index 73ceeff..1a809fc 100644
--- a/conf/cassandra.yaml
+++ b/conf/cassandra.yaml
@@ -676,7 +676,7 @@ native_transport_port: 9042
 # native_transport_max_concurrent_connections_per_ip: -1
 
 # Whether to start the thrift rpc server.
-start_rpc: false
+start_rpc: true
 
 # The address or interface to bind the Thrift RPC service and native transport
 # server to.
